File: _inc/laravel/app/Http/Controllers/Companies/JobCategoryController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    DatabaseConstants,
    UsersConstants,
    ViewsConstants
};
use App\Models\JobCategory;
use App\Traits\ChecksLogin;
use App\Traits\ChecksPermissions;
use Illuminate\Http\{RedirectResponse, Request};
use Illuminate\Support\Facades\{Auth, Log, Route, Validator, View as ViewFacade};
use Illuminate\View\View;

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
class JobCategoryController extends Controller
{
    use ChecksLogin, ChecksPermissions;

    public function index(Request $request): View|RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $action, $method, $class, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if ($c = self::guard($request, 'manage job category', ViewsConstants::JB_CAT . '.index')) return $c;
            Log::debug("[{$base}::{$action}] start", [UsersConstants::COL_USER_ID => $user->id, 'method' => $method]);
            try {
                $qStart = microtime(true);
                $categories = JobCategory::where(DatabaseConstants::COL_TABLE_CREATOR, $user->creatorId())->get();
                $this->logExecutionTime($qStart, $action, 'fetchCategories');
                Log::debug("[{$base}::{$action}] fetched", ['count' => $categories->count()]);
                $viewPath = ViewsConstants::JB_CAT . '.' . $action;
                if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                $renderStart = microtime(true);
                $resp = view($viewPath, compact('categories'));
                $this->logExecutionTime($renderStart, $action, 'renderView');
                return $resp;
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] exception", ['type' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function create(Request $request): View|RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $action, $method, $class, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if ($c = self::guard($request, 'create job category', ViewsConstants::JB_CAT . '.index')) return $c;
            Log::debug("[{$base}::{$action}] start", [UsersConstants::COL_USER_ID => $request->user()?->id, 'method' => $method]);
            try {
                $viewPath = ViewsConstants::JB_CAT . '.' . $action;
                if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                $renderStart = microtime(true);
                $resp = view($viewPath);
                $this->logExecutionTime($renderStart, $action, 'renderView');
                return $resp;
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] exception", ['type' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function show(Request $request, JobCategory $jobCategory): View|RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $jobCategory, $action, $method, $class, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if ($c = self::guard($request, 'view job category', ViewsConstants::JB_CAT . '.index')) return $c;
            Log::debug("[{$base}::{$action}] start", [UsersConstants::COL_USER_ID => $request->user()?->id, 'category_id' => $jobCategory->id, 'method' => $method]);
            try {
                $viewPath = ViewsConstants::JB_CAT . '.' . $action;
                if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                $renderStart = microtime(true);
                $resp = view($viewPath, compact('jobCategory'));
                $this->logExecutionTime($renderStart, $action, 'renderView');
                return $resp;
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] exception", ['type' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($request, $e, $class . '::' . $action, route(ViewsConstants::JB_CAT . '.index'));
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'category_id' => $jobCategory->id]);
    }

    public function store(Request $request): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $action, $class, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if ($c = self::guard($request, 'create job category', ViewsConstants::JB_CAT . '.index')) return $c;
            $v = Validator::make($request->all(), ['title' => 'required']);
            if ($v->fails()) {
                Log::debug("[{$base}::{$action}] validation failed", ['errors' => $v->errors()->all()]);
                return redirect()->back()->with('error', $v->getMessageBag()->first());
            }
            try {
                $crtStart = microtime(true);
                $category = JobCategory::create([
                    'title' => $request->title,
                    DatabaseConstants::COL_TABLE_CREATOR => $request->user()->creatorId()
                ]);
                $this->logExecutionTime($crtStart, $action, 'createJobCategory');
                Log::info("[{$base}::{$action}] created", ['category_id' => $category->id]);
                return redirect()->route(ViewsConstants::JB_CAT . '.index')->with('success', __('Job category successfully created.'));
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] exception", ['type' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function edit(Request $request, string $id): View|RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $id, $action, $method, $class, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if ($c = self::guard($request, 'edit job category', ViewsConstants::JB_CAT . '.index')) return $c;
            Log::debug("[{$base}::{$action}] start", [UsersConstants::COL_USER_ID => $request->user()?->id, 'category_id' => $id, 'method' => $method]);
            try {
                $qStart = microtime(true);
                $jobCategory = JobCategory::where(DatabaseConstants::COL_TABLE_CREATOR, $request->user()->creatorId())->findOrFail($id);
                $this->logExecutionTime($qStart, $action, 'fetchCategory');
                $viewPath = ViewsConstants::JB_CAT . '.' . $action;
                if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                $renderStart = microtime(true);
                $resp = view($viewPath, compact('jobCategory'));
                $this->logExecutionTime($renderStart, $action, 'renderView');
                return $resp;
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage(), 'category_id' => $id]);
                Log::debug("[{$base}::{$action}] exception", ['type' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'category_id' => $id]);
    }

    public function update(Request $request, string $id): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $id, $action, $class, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if ($c = self::guard($request, 'edit job category', ViewsConstants::JB_CAT . '.index')) return $c;
            $v = Validator::make($request->all(), ['title' => 'required']);
            if ($v->fails()) {
                Log::debug("[{$base}::{$action}] validation failed", ['errors' => $v->errors()->all()]);
                return redirect()->back()->with('error', $v->getMessageBag()->first());
            }
            try {
                $qStart = microtime(true);
                $jobCategory = JobCategory::where(DatabaseConstants::COL_TABLE_CREATOR, $request->user()->creatorId())->findOrFail($id);
                $this->logExecutionTime($qStart, $action, 'fetchCategory');
                $updStart = microtime(true);
                $jobCategory->title = $request->title;
                $jobCategory->save();
                $this->logExecutionTime($updStart, $action, 'updateCategory');
                Log::info("[{$base}::{$action}] updated", ['category_id' => $id]);
                return redirect()->route(ViewsConstants::JB_CAT . '.index')->with('success', __('Job category successfully updated.'));
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage(), 'category_id' => $id]);
                Log::debug("[{$base}::{$action}] exception", ['type' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'category_id' => $id]);
    }

    public function destroy(Request $request, string $id): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $id, $action, $method, $class, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if ($c = self::guard($request, 'delete job category', ViewsConstants::JB_CAT . '.index')) return $c;
            Log::debug("[{$base}::{$action}] start", [UsersConstants::COL_USER_ID => $request->user()?->id, 'category_id' => $id, 'method' => $method]);
            try {
                $jobCategory = JobCategory::findOrFail($id);
                if ($jobCategory->created_by !== $request->user()->creatorId()) return defaultPermissionDenial($request, new \Exception('owner'), $class . '::' . $action, route(ViewsConstants::JB_CAT . '.index'));
                $delStart = microtime(true);
                $jobCategory->delete();
                $this->logExecutionTime($delStart, $action, 'deleteCategory');
                Log::info("[{$base}::{$action}] deleted", ['category_id' => $id]);
                return redirect()->route(ViewsConstants::JB_CAT . '.index')->with('success', __('Job category successfully deleted.'));
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage(), 'category_id' => $id]);
                Log::debug("[{$base}::{$action}] exception", ['type' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'category_id' => $id]);
    }

    private function _authorize(Request $request, string $ability): RedirectResponse|null
    {
        if (!$request->user()->can($ability)) {
            Log::warning(__METHOD__ . ' permission denied', [
                UsersConstants::COL_USER_ID => Auth::id(),
                'ability' => $ability
            ]);
            return defaultPermissionDenial(
                $request,
                null,
                __CLASS__ . '::' . __FUNCTION__,
                route(ViewsConstants::JB_CAT . '.index')
            );
        }
        return null;
    }
}

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/companies/job_category_controller.py
import logging
import inspect
from django.core.exceptions import PermissionDenied
from django.contrib import messages
from django.db import transaction
from django.shortcuts import redirect, render
from django.http import HttpRequest, HttpResponse
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from ....Models.companies.job_category import JobCategory
from .._traits.controller import Controller

logger = logging.getLogger(__name__)

class JobCategoryController(Controller):

  @classmethod
  def index(cls, request: HttpRequest) -> HttpResponse:
    CN = cls.__name__
    MN = inspect.currentframe().f_code.co_name
    try:
      if not request.user.has_perm('manage_job_category'):
        return default_permission_denial(
          request,
          err=PermissionDenied('manage_job_category'),
          ref=f'{CN}::{MN}',
          logger=logger
        )
      categories = JobCategory.objects.filter(created_by=request.user.id)
      return render(request, 'jobCategory/index.html', {'categories': categories})
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=f'{CN}::{MN}',
        logger=logger
      )

  @classmethod
  def create(cls, request: HttpRequest) -> HttpResponse:
    CN = cls.__name__
    MN = inspect.currentframe().f_code.co_name
    try:
      return render(request, 'jobCategory/create.html')
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=f'{CN}::{MN}',
        logger=logger
      )

  @classmethod
  def store(cls, request: HttpRequest) -> HttpResponse:
    CN = cls.__name__
    MN = inspect.currentframe().f_code.co_name
    if request.method != 'POST':
      return redirect(get_redirect_url(request))
    try:
      if not request.user.has_perm('create_job_category'):
        return default_permission_denial(
          request,
          err=PermissionDenied('create_job_category'),
          ref=f'{CN}::{MN}',
          logger=logger
        )
      with transaction.atomic():
        title = request.POST.get('title', '').strip()
        if not title:
          messages.error(request, 'Title is required.')
          return redirect(get_redirect_url(request))
        jc = JobCategory(title=title, created_by=request.user.id)
        jc.save()
        messages.success(request, 'Job category successfully created.')
        return redirect(get_redirect_url(request))
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=f'{CN}::{MN}',
        logger=logger
      )

  @classmethod
  def show(cls, request: HttpRequest, jobcategory_id: int) -> HttpResponse:
    CN = cls.__name__
    MN = inspect.currentframe().f_code.co_name
    try:
      return redirect(get_redirect_url(request))
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=f'{CN}::{MN}',
        logger=logger
      )

  @classmethod
  def edit(cls, request: HttpRequest, jobcategory_id: int) -> HttpResponse:
    CN = cls.__name__
    MN = inspect.currentframe().f_code.co_name
    try:
      jc = JobCategory.objects.filter(id=jobcategory_id).first()
      if not jc:
        messages.error(request, 'Job category not found.')
        return redirect(get_redirect_url(request))
      return render(request, 'jobCategory/edit.html', {'jobCategory': jc})
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=f'{CN}::{MN}',
        logger=logger
      )

  @classmethod
  def update(cls, request: HttpRequest, jobcategory_id: int) -> HttpResponse:
    CN = cls.__name__
    MN = inspect.currentframe().f_code.co_name
    if request.method != 'POST':
      return redirect(get_redirect_url(request))
    try:
      if not request.user.has_perm('edit_job_category'):
        return default_permission_denial(
          request,
          err=PermissionDenied('edit_job_category'),
          ref=f'{CN}::{MN}',
          logger=logger
        )
      with transaction.atomic():
        jc = JobCategory.objects.filter(id=jobcategory_id).first()
        if not jc:
          messages.error(request, 'Job category not found.')
          return redirect(get_redirect_url(request))
        title = request.POST.get('title', '').strip()
        if not title:
          messages.error(request, 'Title is required.')
          return redirect(get_redirect_url(request))
        jc.title = title
        jc.save()
        messages.success(request, 'Job category successfully updated.')
        return redirect(get_redirect_url(request))
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=f'{CN}::{MN}',
        logger=logger
      )

  @classmethod
  def destroy(cls, request: HttpRequest, jobcategory_id: int) -> HttpResponse:
    CN = cls.__name__
    MN = inspect.currentframe().f_code.co_name
    try:
      if not request.user.has_perm('delete_job_category'):
        return default_permission_denial(
          request,
          err=PermissionDenied('delete_job_category'),
          ref=f'{CN}::{MN}',
          logger=logger
        )
      with transaction.atomic():
        jc = JobCategory.objects.filter(
          id=jobcategory_id,
          created_by=request.user.id
        ).first()
        if not jc:
          messages.error(request, 'Permission denied.')
          return redirect(get_redirect_url(request))
        jc.delete()
        messages.success(request, 'Job category successfully deleted.')
        return redirect(get_redirect_url(request))
    except Exception as e:
      return default_undefined_exception(
        request,
        err=e,
        ref=f'{CN}::{MN}',
        logger=logger
      )
